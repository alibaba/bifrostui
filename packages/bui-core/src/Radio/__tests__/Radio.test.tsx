import React, { useState } from 'react';
import { act, fireEvent, isConformant, render, screen } from 'testing';
import { Radio, RadioGroup } from '..';

describe('Radio', () => {
  const rootClass = 'bui-radio';

  isConformant({
    className: rootClass,
    displayName: 'BuiRadio',
    Component: Radio,
  });

  it('should be checked', () => {
    const onChange = jest.fn();
    render(
      <Radio checked onChange={onChange}>
        淘票票
      </Radio>,
    );
    const radio = screen.getByRole<HTMLInputElement>('radio');
    expect(radio.checked).toBeTruthy();
  });

  it('should have value', () => {
    render(<Radio value="淘票票" />);
    const radio = screen.getByRole<HTMLInputElement>('radio');
    expect(radio.value).toBe('淘票票');
  });

  it('should have name property', () => {
    render(<Radio name="radio">Radio</Radio>);
    const radio = screen.getByRole<HTMLInputElement>('radio');
    expect(radio.name).toBe('radio');
  });

  it('should be disabled', () => {
    render(<Radio disabled data-testid="radio" />);
    expect(screen.getByTestId('radio')).toHaveClass(`${rootClass}-disabled`);
    expect(screen.getByRole('radio')).toHaveAttribute('disabled');
  });

  it('should use custom unchecked icon', () => {
    const uncheckedIcon = <div className="test-unchecked-icon">icon</div>;
    render(<Radio icon={uncheckedIcon} />);
    const icon = screen.getByText('icon');
    expect(icon).toBeVisible();
  });

  it('should use custom checked icon', () => {
    const onChange = jest.fn();
    const checkedIcon = <div className="test-checked-icon">icon</div>;
    render(<Radio checkedIcon={checkedIcon} checked onChange={onChange} />);
    const icon = screen.getByText('icon');
    expect(icon).toBeVisible();
  });

  describe('Radio label placement', () => {
    const labelPlacements = ['top', 'bottom', 'left', 'right'] as const;
    labelPlacements.forEach((placement) => {
      it(`label should on the ${placement}`, () => {
        const { getByTestId } = render(
          <Radio data-testid={rootClass} labelPlacement={placement}>
            淘票票
          </Radio>,
        );
        const radio = getByTestId(rootClass);
        expect(radio).toMatchSnapshot();
      });
    });
  });

  it('should call `onChange`', () => {
    const fakeChange = jest.fn((e, data) => data.checked);
    const { container } = render(
      <Radio value="淘票票" onChange={fakeChange}>
        淘票票
      </Radio>,
    );
    act(() => {
      const radio = container.querySelector(`.${rootClass}-input`);
      fireEvent.click(radio);
      expect(fakeChange).toReturnWith(true);
    });
  });

  describe('RadioGroup', () => {
    it('value should be controled by `value` property', () => {
      function RadioComponent() {
        const [value] = useState('淘票票');
        return (
          <RadioGroup value={value}>
            <Radio value="水门桥">水门桥</Radio>
            <Radio value="淘票票">淘票票</Radio>
            <Radio value="喜洋洋">喜洋洋</Radio>
          </RadioGroup>
        );
      }

      const { container } = render(<RadioComponent />);
      const radioList = container.querySelectorAll<HTMLInputElement>(
        `.${rootClass}-input`,
      );
      expect(radioList[0].checked).toBeFalsy();
      expect(radioList[1].checked).toBeTruthy();
    });

    it('should call `onChange`', () => {
      const fakeChange = jest.fn((e, data) => data.value);
      const { container } = render(
        <RadioGroup onChange={fakeChange}>
          <Radio value="水门桥">水门桥</Radio>
          <Radio value="淘票票">淘票票</Radio>
          <Radio value="喜洋洋">喜洋洋</Radio>
        </RadioGroup>,
      );
      const radioList = container.querySelectorAll(`.${rootClass}-input`);
      fireEvent.click(radioList[1]);
      expect(fakeChange).toReturnWith('淘票票');
    });

    it('should disabled', () => {
      const { container } = render(
        <RadioGroup disabled>
          <Radio value="水门桥">水门桥</Radio>
          <Radio value="淘票票">淘票票</Radio>
          <Radio value="喜洋洋">喜洋洋</Radio>
        </RadioGroup>,
      );
      const radioList = container.querySelectorAll<HTMLInputElement>(
        `.${rootClass}-input`,
      );
      radioList.forEach((item) => {
        expect(item.disabled).toBeTruthy();
      });
    });
  });

  describe('miniapp Radio', () => {
    const originalModule = jest.requireActual('@bifrostui/utils');
    const restApi = jest.requireActual('react');
    it('should call `onChange`', async () => {
      jest.resetModules();
      jest.doMock('@bifrostui/utils', () => ({
        ...originalModule,
        isMini: true,
      }));
      jest.doMock('react', () => ({
        ...restApi,
      }));
      const { default: MiniRadio } = await import('../index');
      const fakeChange = jest.fn((e, data) => data.checked);
      const { container } = render(
        <MiniRadio value="淘票票" onChange={fakeChange}>
          淘票票
        </MiniRadio>,
      );
      const radio = container.querySelector(`.${rootClass}-input`);
      fireEvent.click(radio);
      expect(fakeChange).toReturnWith(true);
    });
  });
});
