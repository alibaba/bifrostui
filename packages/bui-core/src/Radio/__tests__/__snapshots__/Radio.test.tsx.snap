// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio Radio label placement label should on the bottom 1`] = `
<div
  class="bui-radio bui-radio-label-bottom"
  data-testid="bui-radio"
>
  <input
    class="bui-radio-input"
    type="radio"
    value=""
  />
  <div
    class="bui-radio-icon"
  >
    <span
      class="bui-radio-btn-icon"
    >
      <svg
        class="bui-svg-icon bui-radio-btn-icon-outline icon-size-medium"
        color="#959aa5"
        focusable="false"
        viewBox="0 0 96 96"
      >
        <path
          d="M48 88C25.9 88 8 70.1 8 48S25.9 8 48 8s40 17.9 40 40-17.9 40-40 40zm0-6c18.8 0 34-15.2 34-34S66.8 14 48 14 14 29.2 14 48s15.2 34 34 34z"
        />
      </svg>
      <svg
        class="bui-svg-icon bui-radio-btn-icon-contained icon-size-medium"
        color="#959aa5"
        focusable="false"
        viewBox="0 0 96 96"
      >
        <path
          d="M28 48a20 20 0 1 0 40 0 20 20 0 1 0-40 0"
        />
      </svg>
    </span>
  </div>
  <div
    class="bui-radio-label"
  >
    淘票票
  </div>
</div>
`;

exports[`Radio Radio label placement label should on the left 1`] = `
<div
  class="bui-radio bui-radio-label-left"
  data-testid="bui-radio"
>
  <input
    class="bui-radio-input"
    type="radio"
    value=""
  />
  <div
    class="bui-radio-icon"
  >
    <span
      class="bui-radio-btn-icon"
    >
      <svg
        class="bui-svg-icon bui-radio-btn-icon-outline icon-size-medium"
        color="#959aa5"
        focusable="false"
        viewBox="0 0 96 96"
      >
        <path
          d="M48 88C25.9 88 8 70.1 8 48S25.9 8 48 8s40 17.9 40 40-17.9 40-40 40zm0-6c18.8 0 34-15.2 34-34S66.8 14 48 14 14 29.2 14 48s15.2 34 34 34z"
        />
      </svg>
      <svg
        class="bui-svg-icon bui-radio-btn-icon-contained icon-size-medium"
        color="#959aa5"
        focusable="false"
        viewBox="0 0 96 96"
      >
        <path
          d="M28 48a20 20 0 1 0 40 0 20 20 0 1 0-40 0"
        />
      </svg>
    </span>
  </div>
  <div
    class="bui-radio-label"
  >
    淘票票
  </div>
</div>
`;

exports[`Radio Radio label placement label should on the right 1`] = `
<div
  class="bui-radio bui-radio-label-right"
  data-testid="bui-radio"
>
  <input
    class="bui-radio-input"
    type="radio"
    value=""
  />
  <div
    class="bui-radio-icon"
  >
    <span
      class="bui-radio-btn-icon"
    >
      <svg
        class="bui-svg-icon bui-radio-btn-icon-outline icon-size-medium"
        color="#959aa5"
        focusable="false"
        viewBox="0 0 96 96"
      >
        <path
          d="M48 88C25.9 88 8 70.1 8 48S25.9 8 48 8s40 17.9 40 40-17.9 40-40 40zm0-6c18.8 0 34-15.2 34-34S66.8 14 48 14 14 29.2 14 48s15.2 34 34 34z"
        />
      </svg>
      <svg
        class="bui-svg-icon bui-radio-btn-icon-contained icon-size-medium"
        color="#959aa5"
        focusable="false"
        viewBox="0 0 96 96"
      >
        <path
          d="M28 48a20 20 0 1 0 40 0 20 20 0 1 0-40 0"
        />
      </svg>
    </span>
  </div>
  <div
    class="bui-radio-label"
  >
    淘票票
  </div>
</div>
`;

exports[`Radio Radio label placement label should on the top 1`] = `
<div
  class="bui-radio bui-radio-label-top"
  data-testid="bui-radio"
>
  <input
    class="bui-radio-input"
    type="radio"
    value=""
  />
  <div
    class="bui-radio-icon"
  >
    <span
      class="bui-radio-btn-icon"
    >
      <svg
        class="bui-svg-icon bui-radio-btn-icon-outline icon-size-medium"
        color="#959aa5"
        focusable="false"
        viewBox="0 0 96 96"
      >
        <path
          d="M48 88C25.9 88 8 70.1 8 48S25.9 8 48 8s40 17.9 40 40-17.9 40-40 40zm0-6c18.8 0 34-15.2 34-34S66.8 14 48 14 14 29.2 14 48s15.2 34 34 34z"
        />
      </svg>
      <svg
        class="bui-svg-icon bui-radio-btn-icon-contained icon-size-medium"
        color="#959aa5"
        focusable="false"
        viewBox="0 0 96 96"
      >
        <path
          d="M28 48a20 20 0 1 0 40 0 20 20 0 1 0-40 0"
        />
      </svg>
    </span>
  </div>
  <div
    class="bui-radio-label"
  >
    淘票票
  </div>
</div>
`;
